### PR TITLE
Add a LogManager for Collider which creates a separate log file on each ...

### PR DESCRIPTION
**Pull request by jiayliu**
...day and deletes logs older than 30 days.

### CODE REVIEW

**Comment by jiayliu** [comment]
@milkypostman

File: src/collider/collider/logmanager.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path"
+	"strings"
+	"time"
+)
+
+// LogManager redirects log output to files, one for each |cycle|, and deletes files that are last modified before the retention period.
```
Comment (tkchin) [inline src/collider/collider/logmanager.go]: With the disclaimer that I don't know Go style - this comment seems to exceed 80 chars.

File: src/collider/collider/logmanager.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+func NewLogManager(d string, c time.Duration, r time.Duration) *LogManager {
+	if d == "" {
+		log.Printf("Invalid log dir")
+		return nil
+	}
+	if c >= r {
+		log.Printf("Log retention time must be greater than log cycle")
+		return nil
+	}
+	return &LogManager{
```
Comment (tkchin) [inline src/collider/collider/logmanager.go]: Can you run gofmt if you haven't already? Probably the best way to deal with style in the absence of someone with readability. Thanks.

**Comment by milkypostman** [comment]
I don't think we should build this functionality into the server. We can accomplish the same thing with the following two server changes.

1. run collider with `collidermain >> /path/to/log/file`
2. Add a `collider` file to `/etc/logrotate.d` that contains the following,

```
/path/to/log/file {
    daily
    compress
    copytruncate
    dateext
    missingok
    notifempty
    rotate 10
    sharedscripts
}
```

**Comment by jiayliu** [comment]
I haven't figured out how to rotate the log without restarting Collider. If
logrotate removes the old log and Collider is not restarted, no more
logging will be saved.

On Fri, Feb 20, 2015 at 11:24 AM, Donald Ephraim Curtis <
notifications@github.com> wrote:

> I don't think we should build this functionality into the server. We can
> accomplish the same thing with the following two server changes.
>
>    1. run collider with collidermain >> /path/to/log/file
>    2. Add a collider file to /etc/logrotate.d that contains the following,
>
> /path/to/log/file {
>     daily
>     compress
>     copytruncate
>     dateext
>     missingok
>     notifempty
>     rotate 10
>     sharedscripts
> }
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/webrtc/apprtc/pull/85#issuecomment-75302617>.
>



-- 
Jiayang Liu

**Comment by milkypostman** [comment]
This shouldn't be an issue because we are using the `copytruncate` and using a `>>` redirect from the shell. I am happy to write an application that runs a loop and logs things but from my [limit research](http://unix.stackexchange.com/questions/42544/does-redirecting-output-to-a-file-apply-a-lock-on-the-file) (i see this same advice elsewhere) it won't be an issue. 

In other words, bash/sh will be directing output from `collidermain` to `/path/to/log/file` and when logrotate runs it will copy `file` and truncate it.

The other option is to add a command to logrotate to restart collider in the case of log rotation but I don't think that's necessary and it would wreck some call experience right? Like if people are in a call when collider restarts?

**Comment by jiayliu** [comment]
SG

On Fri, Feb 20, 2015 at 12:02 PM, Donald Ephraim Curtis <
notifications@github.com> wrote:

> This shouldn't be an issue because we are using the copytruncate and
> using a >> redirect from the shell. I am happy to write an application
> that runs a loop and logs things but from my limit research
> <http://unix.stackexchange.com/questions/42544/does-redirecting-output-to-a-file-apply-a-lock-on-the-file>
> (i see this same advice elsewhere) it won't be an issue.
>
> In other words, bash/sh will be directing output from collidermain to
> /path/to/log/file and when logrotate runs it will copy file and truncate
> it.
>
> The other option is to add a command to logrotate to restart collider in
> the case of log rotation but I don't think that's necessary and it would
> wreck some call experience right? Like if people are in a call when
> collider restarts?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/webrtc/apprtc/pull/85#issuecomment-75309196>.
>



-- 
Jiayang Liu

**Comment by jiayliu** [comment]
Let's go with the logrotate solution

**Pull request closed without merging**